src/main/daml/Daml/Finance/Instrument/Swap/Fpml/Util.daml

```diff
@@ -210,12 +211,45 @@ calculateFixPaymentClaimsFromSwapStream fixedRateSchedule s periodicSchedule cal
     claims = if issuerPaysLeg then claimAmounts else give claimAmounts
     claimsTagged = prepareAndTagClaims [claims] "Fix rate payment"
 
-    -- todo: implement principal claims here:
-    --allTaggedClaims = [floatingClaimsTagged, principalClaimsTagged]
     allTaggedClaims = [claimsTagged]
 
   pure allTaggedClaims
 
+-- | Create principal exchange claims
+calculatePrincipalExchangePaymentClaims : [SchedulePeriod] -> Bool -> Deliverable -> Optional Text -> [(Decimal, Bool)] -> [Date] -> PrincipalExchanges -> TaggedClaim
+calculatePrincipalExchangePaymentClaims paymentScheduleAligned issuerPaysLeg currency fxRateId notionals rateFixingDates pe = do
+  let
+    createPrincipalClaim fixingDate p (n, fxAdjRequired) startOrEndDate =
+      when (TimeGte fixingDate) $ scale notional $ when (TimeGte paymentDate) $ one currency
+        where
+          paymentDate = startOrEndDate p
+          notionalConst = Const n
+          notional = case fxRateId of
+            None -> notionalConst
+            Some fxRateId -> if fxAdjRequired then notionalConst * fxRateObs else notionalConst
+              where fxRateObs = Observe fxRateId
+
+    createInitialClaim f p (n, fxAdjRequired) =
+      if issuerPaysLeg then give principalInitialClaim else principalInitialClaim
+        where principalInitialClaim = createPrincipalClaim f p (n, fxAdjRequired) (.adjustedStartDate)
+
+    createIntermediateClaim f p (n, fxAdjRequired) = createPrincipalClaim f p (n, fxAdjRequired) (.adjustedEndDate)
```

Comment:
For my understanding, what is an intermediate claim?

Reply:
It represents an interim exchange of principal during the lifetime of the swap: https://www.fpml.org/spec/fpml-5-11-3-lcwd-1/html/confirmation/schemaDocumentation/schemas/fpml-ird-5-11_xsd/complexTypes/InterestRateStream/principalExchanges.html

Reply:
Thanks, it was indeed also specified in the linked issue 😅 next time I should read that first

Reply:
No problem, thanks for the review! :)